CONTRIBUTING.md

```diff
@@ -78,96 +85,160 @@ installed.
+is considered an expert task. To find out how to do it please read
+`.github/workflows/test.yml` and replicate it for your local setup.
+
+After working on BinderHub you can [clean up](#cleaning-up-resources).
+
+
+### One-time installation
+
+To run a full BinderHub it needs to have access to a kubernetes cluster
+with a JupyterHub installed on it. This is what we will setup in this section.
+All the steps are given as command-line commands for Ubuntu systems. They are
+used as a common denominator that can be translated into the correct commands
+on your local system.
+
+Before you begin, there are a few utilities that need to be installed:
+```bash
+sudo apt install python3 python3-pip npm curl
+```
+
+If you are on linux, you may additionally need to install socat for port forwarding:
+
+```bash
+sudo apt install socat
+```
+
+1. [Install Minikube](https://minikube.sigs.k8s.io/docs/start/) to run Kubernetes locally.
+
+  To start your minikube cluster , run the command: `minikube start`. This
+  starts a local kubernetes cluster inside a virtual machine. This command
+  assumes that you have already installed one of the VM drivers: virtualbox,
+  xhyve or KVM2.
+
+1. [Install helm](https://helm.sh/docs/intro/quickstart/#install-helm) to
+   install JupyterHub and BinderHub on your cluster.
+
+1. Add the JupyterHub helm charts repo:
+
+  ```bash
+  helm repo add jupyterhub https://jupyterhub.github.io/helm-chart/
+  helm repo update
+  ```
+
+1. Install BinderHub the Python package and its development requirements:
 
    ```bash
    python3 -m pip install -e . -r dev-requirements.txt
    ```
 
-1. Install the JupyterHub Helm chart by itself into your Kubernetes cluster
-   current namespace.
+1. Install JupyterHub in minikube with helm
 
-   ```bash
-   # Append --auth here if you want to develop against a non-public BinderHub
-   # that relies on JupyterHub's configured Authenticator to decide if the users
-   # are allowed access or not.
-   ./testing/local-binder-k8s-hub/install-jupyterhub-chart
-   ```
+  ```bash
+  ./testing/local-binder-k8s-hub/install-jupyterhub-chart
+  ```
 
-1. Configure `docker` using environment variables to use the same Docker daemon
-   as your `minikube` cluster. This means images you build are directly
-   available to the cluster.
+1. Before starting your local deployment run:
 
-   ```bash
-   eval $(minikube docker-env)
-   ```
+  ```bash
+  eval $(minikube docker-env)
+  ```
 
-1. Start BinderHub with the testing config file.
+  This command sets up `docker` to use the same Docker daemon as your minikube
+  cluster does. This means images you build are directly available to the
+  cluster. Note: when you no longer wish to use the minikube host, you can
+  undo this change by running:
 
-   ```bash
-   python3 -m binderhub -f testing/local-binder-k8s-hub/binderhub_config.py
-   ```
+  ```bash
+  eval $(minikube docker-env -u)
+  ```
```

Comment:
3 step indentation instead of 2 to let the text become nested under the numerical bullet point.

Related to this, I've been very happy using prettier to format markdown in z2jh as introduced by Simon - it fixed all those boring details.

Reply:
Do you know how I can run prettier? Is it a command-line tool like `black`?

Reply:
Yepp, check out the z2jh pre-commit config for example. I have never run it manually myself but instead relied on pre-commit to do it.

Reply:
`pre-commit` is the easiest way to run it, even if you want to run it manually. It's written in JavaScript so if you install it manually you'll have to deal with npm, whereas pre-commit will magically hide all that:
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/9dcf8f80c0d2f8a5d1b4b7d0ebddf0896e84e453/.pre-commit-config.yaml#L35-L39 (https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1996)